### PR TITLE
[WIP] Multiple hosts 98

### PR DESCRIPTION
**Pull request by Rattenkrieg**
I'm using vert.x pg client in my cruds and want to establish two pools: one connected to primary server for oltp workloads and second one connected to multiple secondary servers for read only (analytical) queries. Seems there is already some demand, see #98

### CODE REVIEW

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/SqlConnectOptions.java [hunk 1]
```diff
@@ -58,60 +55,96 @@ public SqlConnectOptions(JsonObject json) {
...
 
   /**
+   * @deprecated use {@link SqlConnectOptions#getHosts()}
+   *
    * Get the host for connecting to the server.
    *
    * @return the host
    */
+  @Deprecated
+  @GenIgnore
```
Comment (Rattenkrieg) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/SqlConnectOptions.java:81]: Ignored to make json converters use `getHosts()`
Setters are not ignored due to backward compatibility reasons, i.e. to construct instance of options from json generated from old version.

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/SqlConnectOptions.java [hunk 2]
```diff
@@ -58,60 +55,96 @@ public SqlConnectOptions(JsonObject json) {
...
   }
 
   /**
+   * @deprecated use {@link SqlConnectOptions#getHosts()}
+   *
    * Get the host for connecting to the server.
    *
    * @return the host
    */
+  @Deprecated
```
Comment (Rattenkrieg) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/SqlConnectOptions.java:80]: Deprecation decision is up to maintainers.

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/SqlConnectOptions.java [hunk 3]
```diff
@@ -287,9 +320,46 @@ public SqlConnectOptions addProperty(String key, String value) {
...
+  public List<SqlHost> getHosts() {
+    ArrayList<SqlHost> hosts = new ArrayList<>(this.hosts.size());
+    for (SqlHost host : this.hosts) {
+      hosts.add(new SqlHost(host));
+    }
+    return hosts;
+  }
+
+  @GenIgnore
+  public SqlConnectOptions setHosts(SqlHost... hosts) {
```
Comment (Rattenkrieg) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/SqlConnectOptions.java:340]: Api shape decision is up to maintainers.

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SqlConnectionFactoryBase.java [hunk 4]
```diff
@@ -75,35 +80,79 @@ protected SqlConnectionFactoryBase(VertxInternal vertx, SqlConnectOptions option
     this.netClient = vertx.createNetClient(netClientOptions, clientCloseFuture);
   }
 
+  private int advanceNextHost() {
+    return lastLiveHost.getAndIncrement() % socketAddresses.size();
+  }
+
+  private static class AddressesIterator implements Iterator<SocketAddress> {
```
Comment (Rattenkrieg) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SqlConnectionFactoryBase.java:87]: Could be enhanced in future to perform iteration based on specific order i.e. emulate pg jdbc `?targetServerType=preferSlave`

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SqlConnectionFactoryBase.java [hunk 5]
```diff
@@ -75,35 +80,79 @@ protected SqlConnectionFactoryBase(VertxInternal vertx, SqlConnectOptions option
...
     ContextInternal context = promise.context();
-    Promise<Connection> promise0 = context.promise();
+    PromiseInternal<Connection> promise0 = context.promise();
     promise0.future().onComplete(ar -> {
       if (ar.succeeded()) {
         promise.complete(ar.result());
       } else {
-        if (remainingAttempts >= 0) {
+        if (addresses.hasNext()) {
+          doConnectWithRetry(promise, remainingAttempts, addresses);
```
Comment (Rattenkrieg) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SqlConnectionFactoryBase.java:145]: This is the gist of this PR. I decided to keep this logic in current method to not to create additional promises.

**Comment by vietj** [comment]
I'm currently implementing this feature in a branch https://github.com/eclipse-vertx/vertx-sql-client/tree/pool-hook

**Comment by vietj** [comment]
you can review it when it will be completed

**Comment by vietj** [comment]
I'll reuse tests from this PR

**Comment by Rattenkrieg** [comment]
Great news @vietj will be happy to review!
Superseded by https://github.com/eclipse-vertx/vertx-sql-client/tree/pool-hook

**Pull request closed without merging**